### ISSUE TITLE
oracledb: commit is a no-op

### ISSUE THREAD

**Issue by jakecoffman** [knex/knex#3612]
# Environment

Knex version: 0.20.6
Database + version: Oracle 12c
OS: Linux & OSX

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

The oracledb commit is a no-op. I think it should perform the actual commit immediately.

2. Error message

There's no error message.

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to MySQL or PostgreSQL, then single file example which initializes
   needed data and demonstrates the problem.

I noticed this in our production code that the commit seemed to not work immediately. The integration tests also seem to exhibit this behavior which is why it may fail sometimes (race conditions.)

Here's the code path it takes:

1. Transaction starts and it sets [isTransaction=true](https://github.com/knex/knex/blob/aea29d0a7c00a4e6e0f6eafdf57240ef211813cb/lib/dialects/oracledb/transaction.js#L60)
2. When commit is called it calls `commitAsync` which [immediately resolves doing nothing](https://github.com/knex/knex/blob/aea29d0a7c00a4e6e0f6eafdf57240ef211813cb/lib/dialects/oracledb/index.js#L112)
3. Later it sets [isTransaction=false](https://github.com/knex/knex/blob/aea29d0a7c00a4e6e0f6eafdf57240ef211813cb/lib/dialects/oracledb/transaction.js#L72) so it really does the commit this time.

The problem is in the tests and in my production code I assume the commit actually happened when commit was called, so we occasionally see errors due to the race condition it creates.

**Comment by jakecoffman** [comment]
I can't reproduce with a test. Starting to wonder if it's not an issue at all and I fixed the main problem in #3606.

Will reopen if it comes up again.

**Comment by akaraivanov** [comment]
I've got some tests of our code that fail once in a while using knex with oracledb dialect. The description of this tickets is a pretty good match to our use case. In short, we create a transaction, insert data, use explicit commit on the transaction, await on the commit then read data in second transaction. When our test fails, the second transaction does not see the data committed by the first.

The code merged by this issue does not really address the description of the issue, but the fix proposed in [2b290e9](https://github.com/jakecoffman/knex/commit/2b290e935cc4190eaa1ad49512592ecf866b7bf0) does.

To create a minimal test case that demonstrate the problem I added following test case to `test/integration/builder/transaction.js` (as part of the Transactions test):

    it('ticket3612', async function() {
      let id = null;
      const trx1 = await knex.transaction();
      const trx2 = await knex.transaction();
      const email = `${Math.random()}@example.com`;
      await trx1('accounts')
            .returning('id')
            .insert({
              first_name: 'foo1',
              last_name: 'foo1',
              email,
              logins: 1,
              about: 'Lorem ipsum Dolore labore incididunt enim.',
              created_at: new Date(),
              updated_at: new Date(),
            });
      await trx1.commit();        
      const rows = await trx2('accounts')
           .select("email")
           .where("email", email);
      await trx2.commit();
      expect(rows).to.have.length(1);
    });

That on its own does not make the test fail, but with the following extra change it fails every time:
```
diff --git a/lib/dialects/oracledb/transaction.js b/lib/dialects/oracledb/transaction.js
index a759470..11c0afd 100644
--- a/lib/dialects/oracledb/transaction.js
+++ b/lib/dialects/oracledb/transaction.js
@@ -70,6 +70,7 @@ module.exports = class Oracle_Transaction extends Transaction {
         return await cb(connection);
       } finally {
         debugTx('%s: releasing connection', this.txid);
+       await Bluebird.delay(1200);
         connection.isTransaction = false;
         try {
           await connection.commitAsync();
```
Next, by applying @jakecoffman's  [2b290e9](https://github.com/jakecoffman/knex/commit/2b290e935cc4190eaa1ad49512592ecf866b7bf0), test case worked again. (I patched only `lib/dialects/oracledb/index.js` because the 2nd changed file `lib/dialects/oracledb/transaction.js` was not changing any semantics).

What do you guys think? Could we reopen this issue, and merge @jakecoffman's [2b290e9](https://github.com/jakecoffman/knex/commit/2b290e935cc4190eaa1ad49512592ecf866b7bf0) fix?

**Comment by jakecoffman** [comment]
I wouldn't mind getting the fix in since it just seems more correct to have `commit` actually perform the commit. We just can't make a test that recreates the problem. Thoughts @kibertoad?

**Comment by akaraivanov** [comment]
There are also 2 integration tests from Transactions group that are skipped for oracledb. I wonder if 2b290e9 would make them stable...

**Comment by kibertoad** [comment]
@akaraivanov If you would like to take a lead on this and contribute test that fails without a fix, I'm all for reopening the issue.

**Comment by akaraivanov** [comment]
I can have a look. Integration-stress test and unit test using sinon spies are doable. Anything else - probably not so much, unless code under test is also changed to provoke a failure as I do above by adding a delay.

**Comment by kibertoad** [comment]
Yeah, should be good enough. Thanks!

**Comment by elhigu** [comment]
This thing also happens every now and then in oracledb migration tests, when there is IIRC wrong lock value left migration lock table or something like that.

**Comment by kibertoad** [comment]
@akaraivanov Released in 0.20.10.

**Comment by kibertoad** [comment]
@akaraivanov Would appreciate it if you could test new version with your use-cases and report back if you'll observe any changes in stability of your tests.

**Comment by akaraivanov** [comment]
@kibertoad , @jakecoffman Thanks for merging the fix and making a release.I will test and let you know.

**Comment by akaraivanov** [comment]
@kibertoad, we run tests (including stress tests) with new version and there have been no problems with that.

**Comment by kibertoad** [comment]
Released in 0.20.11
